Update dynamic instrumentation docs - log probes (#16905)
DataDog/documentation: content/en/dynamic_instrumentation/how-it-works.md

## Overview

Dynamic instrumentation allows you to add probes to your production systems at any location in your application's code, including third-party libraries. Dynamic Instrumentation has low overhead and is guaranteed to have no side effects on your system.

## Probe types

You can create *log probes* and *metric probes*.

### Log probes

> Log probes are enabled by default on all service instances that match the desired environment and version. They are rate limited to 5000 logs/s.

Reviewer: I would also add a line with example on to write a template message: 
"User {user.id} completed transaction with {count(products)} products"

And also mention that conditions can be used to create logs on specific cases an maybe give an example of condition.

Author: :+1: Good call, added this example and a paragraph about conditions on this page. I didn't put an example for a condition in, we have one in the screenshot on the main page.

    - The first 255 characters for string values.
    - 20 fields inside objects. Static fields are not collected.
  - Call **stack trace**.
  - Caught and uncaught **exceptions**.

Since capturing snapshots is a more performance intensive operation, by default it is only enabled on one instance of your service that matches the desired environment and version. Probes with snapshot capture are rate limited to execute once every second.

**Note**: The capture limits are configurable and subject to change while Dynamic Instrumentation is in beta.

### Metric probes

Metric probes are enabled by default on all service instances that match the desired environment and version. Metric probes are not rate limited and execute every time the method or line is invoked.

Dynamic Instrumentation metric probes support the following metric types:

- [**Count**][1]: Counts how many times a given method or line is executed. Can be combined with [metric expressions](#expression-language) to use the value of a variable to increment the count.
- [**Gauge**][2]: Generates a gauge based on the last value of a variable. This metric requires a [metric expression](#expression-language).

> - [**Histogram**][3]: Generates a statistical distribution of a variable. This metric requires a [metric expression](#expression-language).

Author: It's not exposed in the UI, we only show the metric types we list here.

Dotnet doesn't support metric probes at all now.

## Overview

Dynamic instrumentation allows you to add probes to your production systems at any location in your application's code, including third-party libraries. Dynamic Instrumentation has low overhead and is guaranteed to have no side effects on your system.

## Probe types

You can create *log probes* and *metric probes*.

### Log probes

> Log probes are enabled by default on all service instances that match the desired environment and version. They are rate limited to execute at most 5000 times per second, on each service instance.

Reviewer 2: ```suggestion
Log probes are enabled by default on all service instances that match the specified environment and version. They are rate limited to execute at most 5000 times per second, on each service instance.
```

If you enable `Capture method parameters and local variables` on the log probe, Dynamic Instrumentation captures the following data and adds it to the log event:
  - **Method arguments**, **local variables**, and **fields**, with the following limits by default:
    - Follow references three levels deep (configurable in the UI).
    - The first 100 items inside collections.
    - The first 255 characters for string values.
    - 20 fields inside objects. Static fields are not collected.
  - Call **stack trace**.
  - Caught and uncaught **exceptions**.

> Since capturing extra data is a more performance intensive operation, by default it is only enabled on one instance of your service that matches the desired environment and version. Probes with snapshot capture are rate limited to execute once every second.

Reviewer 2: ```suggestion
Because capturing extra data is performance intensive, by default it is enabled on only one instance of your service that matches the specified environment and version. Probes with this capture setting enabled are rate limited to execute once per second.
```

> It's required to set a log message template on every log probe. The template supports embedding expressions inside curly brackets. For example: `User {user.id} purchased {count(products)} products`.

Reviewer 2: ```suggestion
You must set a log message template on every log probe. The template supports embedding expressions inside curly brackets. For example: `User {user.id} purchased {count(products)} products`.
```

> You can also set a condition on a log probe using the [expression language](#expression-language). The expression must evaluate to a boolean, the probe will execute if the expression is true and will not capture or emit any data if the expression is false.

Reviewer 2: ```suggestion
You can also set a condition on a log probe using the [expression language](#expression-language). The expression must evaluate to a Boolean. The probe executes if the expression is true, and does not capture or emit any data if the expression is false.
```

**Note**: The capture limits are configurable and subject to change while Dynamic Instrumentation is in beta.

### Metric probes

> Metric probes are enabled by default on all service instances that match the desired environment and version. Metric probes are not rate limited and execute every time the method or line is invoked.

Reviewer 2: ```suggestion
Metric probes are enabled by default on all service instances that match the specified environment and version. Metric probes are not rate limited and execute every time the method or line is invoked.
```

Dynamic Instrumentation metric probes support the following metric types:

- [**Count**][1]: Counts how many times a given method or line is executed. Can be combined with [metric expressions](#expression-language) to use the value of a variable to increment the count.
- [**Gauge**][2]: Generates a gauge based on the last value of a variable. This metric requires a [metric expression](#expression-language).
- [**Histogram**][3]: Generates a statistical distribution of a variable. This metric requires a [metric expression](#expression-language).

#### Expression language

> The dynamic instrumentation expression language is used in log message templates, metric expressions and probe conditions.

Reviewer 2: ```suggestion
Use the Dynamic Instrumentation expression language in log message templates, metric expressions, and probe conditions.
```

For example, you can create a histogram from the size of a collection using `count(myCollection)` as the metric expression. Metric expressions must evaluate to a number.

In log templates, expressions are delimited from the static parts of the template with brackets, for example: `User name {user.name}`. Log template expressions can evaluate to any value. If evaluating the expression fails, it is replaced with `UNDEFINED`.

> Probe conditions must evaluate to a boolean, for example: `startsWith(user.name, "abc")`, `len(str) > 20` or `a == b`.

Reviewer 2: ```suggestion
Probe conditions must evaluate to a Boolean, for example: `startsWith(user.name, "abc")`, `len(str) > 20` or `a == b`.
```
